source/common/stats/thread_local_store.cc

```diff
@@ -214,19 +263,98 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
   }
 
   std::unique_lock<std::mutex> lock(parent_.lock_);
-  HistogramSharedPtr& central_ref = central_cache_.histograms_[final_name];
+  ParentHistogramImplSharedPtr& central_ref = central_cache_.histograms_[final_name];
+
+  std::vector<Tag> tags;
+  std::string tag_extracted_name = parent_.getTagsForName(final_name, tags);
   if (!central_ref) {
-    std::vector<Tag> tags;
-    std::string tag_extracted_name = parent_.getTagsForName(final_name, tags);
-    central_ref.reset(
-        new HistogramImpl(final_name, parent_, std::move(tag_extracted_name), std::move(tags)));
+    // Since MetricImpl only has move constructor, we are explicitly copying here.
+    std::string central_tag_extracted_name(tag_extracted_name);
+    std::vector<Tag> central_tags(tags);
+    central_ref.reset(new ParentHistogramImpl(
+        final_name, parent_, std::move(central_tag_extracted_name), std::move(central_tags)));
   }
+  TlsHistogramSharedPtr hist_tls_ptr(new ThreadLocalHistogramImpl(
+      final_name, parent_, std::move(tag_extracted_name), std::move(tags)));
+  central_ref->addTlsHistogram(hist_tls_ptr);
 
   if (tls_ref) {
-    *tls_ref = central_ref;
+    *tls_ref = hist_tls_ptr;
   }
+  return *hist_tls_ptr;
+}
 
-  return *central_ref;
+ThreadLocalHistogramImpl::ThreadLocalHistogramImpl(const std::string& name, Store& parent,
+                                                   std::string&& tag_extracted_name,
+                                                   std::vector<Tag>&& tags)
+    : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent),
+      current_active_(0), flags_(0) {
+  histograms_[0] = hist_alloc();
+  histograms_[1] = hist_alloc();
+}
+
+ThreadLocalHistogramImpl::~ThreadLocalHistogramImpl() {
+  hist_free(histograms_[0]);
+  hist_free(histograms_[1]);
+}
+
+void ThreadLocalHistogramImpl::recordValue(uint64_t value) {
+  hist_insert_intscale(histograms_[current_active_], value, 0, 1);
```

Comment:
for each histogram created we have ThreadLocal equivalent that is updated in that thread and and ParentHistogram that holds the summary across all the threads. So I think it is not an issue.

Reply:
Sorry, I think that's exactly my concern.  The thread from which you call `Histogram& hist = some_scope.histogram(some_name);` is assumed to be the same thread that will ultimately call `hist.recordValue()`.  This assumption doesn't generally hold for stats.  Many stats are created in the main thread and then passed around to other threads to modify their values, since all the other stats objects are completely thread safe.  See this stats struct as an example: https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_impl.h#L405.  It is created in the `HttpConnectionManagerConfig`, which is initialized in the main thread.  Then it is passed to all of the `HttpConnectionManager`s that are created for each request (can be on any worker thread).  The worker threads are able to freely mutate the stats IIUC.

Reply:
Hmm, yes. Great catch @mrice32. This is broken, at least for how we commonly use histograms. I think the best option to fix is probably to always return back a parent/main thread histogram, and then have recordValue() internally redirect to the TLS version (and create it if needed). I can have a look more later about other options.

Reply:
Alright, I looked at the code again. Here is my suggestion on how to fix this (though I'm going to admit that this change is complicated enough I'm having trouble doing the design without also being able to try some stuff out via coding, so if it comes to that I can help out). It's going to require some code-rejiggering, but I don't think it's going to end up being too terribly difficult, and it will leave us in a cleaner/better place.

1) Remove the `recordValue()` function from the `Histogram` interface.
2) Replace

Reply:
Alright, I looked at the code again. Here is my suggestion on how to fix this (though I'm going to admit that this change is complicated enough I'm having trouble doing the design without also being able to try some stuff out via coding, so if it comes to that I can help out). It's going to require some code-rejiggering, but I don't think it's going to end up being too terribly difficult, and it will leave us in a cleaner/better place.

1) When you return a Histogram from the thread local store, what you will actually return effectively is the parent histogram, not the TLS histogram.
2) When recordValue() is called on the parent, at that point, you will do the TLS lookup/cache add, and then call recordValue() on the actual TLS version.

This should be pretty easily doable by just changing the flow of the histogram acquisition function that you have now.

Reply:
Very cool; great catch @mrice32 and quick prescription for a remedy @mattklein123 .

Would this have been caught in any load-test we have in place?

Reply:
Other than running at Lyft and crashing? Probably not. :) Would be great to figure out a test that would catch this case. I think one should be able to be crafted pretty easily that TSAN should fail on...

Reply:
@mattklein123 SGTM, I like that solution. I was trying to think of a way to do it without doing a cache lookup for every `recordValue()` call, but I don't think it would be possible without a TLS slot per histogram, and TLS isn't really built for that sort of scaling IIUC (and cache lookups are pretty cheap).

@jmarantz Not sure.  I'm a surprised that tsan didn't get upset about this in one of the existing integration tests.

Reply:
> I was trying to think of a way to do it without doing a cache lookup for every recordValue() call

I think in the future we could add some type of "direct TLS access" histogram for use cases in which the user knows they are safe, but IMO I wouldn't worry about it for now and we can track as a future perf improvement...

Reply:
@ramaraochavali one other thing we should do here is keep track of the thread ID that the TLS histograms are created on and make sure that recordValue() is only called on that same thread. You could add ASSERTs along those lines pretty easily, and it would also help document the threading model.

Reply:
@mattklein123 @mrice32  I have added a tlsHistogram method in the ScopeImpl that creates ThreadLocalHistogram since the ScopeImpl has all the things needed for putting in scope cache etc. It is working fine. Can you PTAL? I think we should be good with this.

@mattklein123  I have added the ASSERT that you mentioned.

Reply:
@mrice32 are you fine with this change?